Link delegation explainer - review updates (#1111)
openui/open-ui: site/src/pages/components/link-area-delegation-explainer.mdx

---
menu: Active Proposals
name: Link Area Delegation (Explainer)

> ##path: /components/...

Author: per commit message, is this necessary?

##pathToResearch: /components/...
layout: ../../layouts/ComponentLayout.astro
---

- Authors and contributors: [@noamr](https://github.com/noamr), [@scottaohara](https://github.com/scottaohara), [@nmn](https://github.com/nmn), [@jakearchibald](https://github.com/jakearchibald)
- Last updated: October 16, 2024
- Feedback: https://github.com/openui/open-ui/issues/

{/* START doctoc generated TOC please keep comment here to allow auto update */}
{/* DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE */}

## State of this explainer
We are currently incubating and exploring a new idea. Chime in!

## Overview

> Historically and for very good reasons, interactive elements on the web cannot be nested within other interactive elements, including hyperlinks.

Reviewer: This sentence is a bit awkward to read. Is this better?

```suggestion
Historically and for very good reasons, interactive elements on the web cannot be nested within other interactive elements. Examples of interactive elements include links and buttons.
```

> Interactive elements not allowing descendent interactive elements is a well established rule in HTML for various interactive elements - where descendent interactive elements or elements with a `tabindex` attribute are specifically called out as invalid (see [the `a` element's content model](https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-a-element). Additionally, the [ARIA in HTML guideline]([https://w3c.github.io/html-aria/](https://w3c.github.io/html-aria/#allowed-descendants-of-aria-roles)) specification provides further guidance to discourage the use of interactive elements as descendent of ARIA roles which are themselves considered "widget" (interactive) roles.

Reviewer: ```suggestion
This restriction on nested interactive elements is a well established rule in HTML - where descendent interactive elements or elements with a `tabindex` attribute within other interactive elements are specifically called out as invalid (see [the `a` element's content model](https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-a-element). Additionally, the [ARIA in HTML guideline]([https://w3c.github.io/html-aria/](https://w3c.github.io/html-aria/#allowed-descendants-of-aria-roles)) further discourages the use of interactive elements as descendent of ARIA roles which are themselves considered "widget" (interactive) roles.
```

Author: thanks.  going to rework this a bit further in an updated commit i'm working on

> Despite this long-standing rule, there are websites which ignore it. In doing so, these sites end up with UI which will be difficult, if not blocking, for people using assistive technology, such as screen readers. On the flip-side, there are also websites such as [Walmart](https://walmart.ca), [Reddit](https://reddit.com), [Rotten Tomatoes](https://rottentomatoes.com), to name a few prominent ones, that use specific markup patterns and CSS or JS to present content that visually appears to have nested targets, as the larger container appears to be clickable - but inspecting the markup patterns would show otherwise.

Reviewer: ```suggestion
Despite this long-standing rule, there are websites which ignore it. In doing so, these sites end up with UI which will be difficult, if not unusable, for people using assistive technology such as screen readers. On the flip-side, there are websites that use CSS and/or JS creatively to create UI that *appears* to have nested interactive elements while avoiding nesting them in the markup itself in order to keep the UI accessible. Examples of such websites include [Walmart](https://walmart.ca), [Reddit](https://reddit.com), and [Rotten Tomatoes](https://rottentomatoes.com) among others.
```

Reviewer: Do we need a more detailed write for the "motivation" for this feature?

Author: Probably should. But I didn’t have time to get into that. That was another one of my comments that I forgot to repost from my review that didn’t get posted since the explainer got merged while I was reviewing

Author: re-reading, i realize maybe you meant that comment just in context to this section.  i'm pulling in some of your ideas / adjusting further.

i'm also changing out the rotten tomatoes example for amazon and best buy, since i could easily find this pattern on both those sites, and still wasn't quite sure where on rotten tomatoes to look.

## The clikcable container or "card" pattern
Commonly a clickable container, often referred to as a "card", will present a graphic or other media, text, a primary call to action (link or button) as well as other supplementary interactive elements:

<figure>
  <img src="https://github.com/user-attachments/assets/27e5076d-bc98-4fb8-a429-e8462ff9024c" 
       width="400" 
       alt="Screenshot of a Reddit topic 'card' from the r/cats subreddit. Shows time of publication, r/cats as a hyperlink, a join and more options button, followed by the title of the post and a graphic of a cat pearing out a window while its owner takes its picture. up/down vote, comment, award and share buttons follow the graphic.">

  <figcaption>
    <p>With this Reddit example, the whole card is clickable and doing so and would navigate a user to the Reddit post. But there are also other interactive elements (up/down vote, comments, join, the sub-reddit link, etc.) are also clickable. These appear to be "nested" within the clickable card, but due to work arounds, this is not the case.</p>
  </figcaption>
</figure>

### Existing ways to achieve clickable containers

> Today, making a container/card clickable can be achieved by a few different approaches, each with its own tradeoffs:

Reviewer: ```suggestion
Today, making a container/card clickable can be achieved with a few different approaches, each with its own tradeoffs:
```

1. Using the `a` element to wrap the content that needs to be clickable.  The issue with this is:
  a. doesn't support nested interactive elements without creating accessibility issues.
  b. using an `a` element to wrap content is not supported in certain contexts - e.g., such as trying to wrap a row within a table.  The HTML parser will remove such hyperlinks from the rendered table markup.

>   c. even without nested interactive elements, the accessible name of the hyperlink can become overly verbose, or may not be exposed as all depending on the element the hyperlink wraps.

Reviewer: ```suggestion
  c. even without nested interactive elements, the accessible label of the hyperlink can become overly verbose, or may not be exposed as all depending on the children elements.
```

Author: Name is the correct term. 
I will revise the other bit as neither of our wording is quite right for that

2. Using CSS to position / stack interactive elements. For instance, using CSS to put the link, or a link's `::before` or `::after` pseudo element "behind" the "nested" interactive elements, so that the link catches clicks that are not handled by these other interactive targets. The issue here is:

>   a. no standardized way to do this, with some implementations better than others - particularly such implementations that use duplicate links, creating redundancy for keyboard and/or screen reader users (even if the duplicate link is `tabindex=-1`).

Reviewer: Does expanding the `::before` or `::after` element create an additional tabbable element? Shouldn't it simply expand the hit area of the existing link or button?

Author: No those don’t receive focus.  This item isn’t talking about that though, but rather implementations that use two actual links. Walmart does this and it’s rather sloppy

Author: updating this to clarify, as i can see why you thought i was saying the pseudo elements would receive focus.

Reviewer: Thanks!

>   b. this sort of pattern often makes it difficult if not impossible for some users to select what appears to be static text. 

Reviewer: ```suggestion
  b. this sort of pattern often makes it difficult if not impossible for some users to select what appears to be static text. 
  c. Such a pattern usually relies on all intermediate elements between the container/card and the primary action within to have `display: static`. Depending on the visual design requirements, this may not be feasible.
```

3. Using JavaScript to Capture clicks: have the container element capture clicks, and delegate them to the main link only if the original target is not another interactive target.
   a. requires the use of JS, which in itself is not a bad thing, but may undesired by some developers - leading them to use alternate patterns, or,
   b. can lead to other UX issues, such as activation on `mousedown` rather than 'up', or people forgoing the inclusion of a nested hyperlink in favor of only using the JS, leading to the element being accessible to pointer events, but often not to keyboard or assistive techology users. 

> 

Reviewer: ```suggestion
  c. can conflict with certain browser behaviours such as context menus and popup blocking.

```

Author: thanks.  rewording slightly but pulling this in

However, something about this feels limiting, as it is constrained to click areas in advance, and doesn't lend itself to future UA enhancements of this experience, such as context menus. Identifying the hyperlink or interactive control to expand its click area might also become difficult if needing to apply this feature to markup beyond one's immediate control - e.g., syndicated content created by other product teams. 

### Approach 4: Use invokers 
An interesting alternative from @jaffathecake is to use [invokers](https://open-ui.org/components/invokers.explainer/). 
The prior "card" example has been modified to use `commandfor` and `commands` attributes, instead of a `linkarea` attribute or CSS:
```html
<section class="card" commandfor="card123-title" commands="click contextmenu">
  <a href="/post?id=123" id="card123-title">Post Title</a>
  <img ...>
  <button>Join</button>
  <button>Share</button>
</section>
```

This a nice idea as it could allow delegation of clicks to more than just "links". For instance, delegating to a `button` element instead. 

> The tradeoff here might be that this would need two attributes to declare, rather than just one. And that the `id` of the desired element to delegate to might be necessary to declare more often than the common case of extending the click-area of the first hyperlink within a container.

Reviewer: As a point of discussion, I think the need for explicit IDs is a positive.

Despite this long-standing rule, there are websites which ignore it. In doing so, these sites end up with UI which will be difficult, if not blocking, for people using assistive technology, such as screen readers. On the flip-side, there are also websites such as [Walmart](https://walmart.ca), [Reddit](https://reddit.com), [Rotten Tomatoes](https://rottentomatoes.com), to name a few prominent ones, that use specific markup patterns and CSS or JS to present content that visually appears to have nested targets, as the larger container appears to be clickable - but inspecting the markup patterns would show otherwise.

## The clikcable container or "card" pattern
Commonly a clickable container, often referred to as a "card", will present a graphic or other media, text, a primary call to action (link or button) as well as other supplementary interactive elements:

<figure>
  <img src="https://github.com/user-attachments/assets/27e5076d-bc98-4fb8-a429-e8462ff9024c" 
       width="400" 

>        alt="Screenshot of a Reddit topic 'card' from the r/cats subreddit. Shows time of publication, r/cats as a hyperlink, a join and more options button, followed by the title of the post and a graphic of a cat pearing out a window while its owner takes its picture. up/down vote, comment, award and share buttons follow the graphic.">

Author: ```suggestion
       alt="Screenshot of a Reddit topic 'card' from the r/cats subreddit. Shows time of publication, r/cats as a hyperlink, a join and more options button, followed by the title of the post and a graphic of a cat peering out a window while its owner takes its picture. up/down vote, comment, award and share buttons follow the graphic." />
```
